### DISCUSSION TITLE
How to backup, restore and migrate

### DISCUSSION THREAD

**Discussion by unknwon** [gogs/gogs#6876] [Show and tell]
Other than pack up `gogs-repositories`, `data`, `custom` and database separately, Gogs provides two commands for unified process of _backup_, _restore_ and even _migrate to another database engine_, because the backup format is database-agnostic.

## Prerequisite

- Gogs version >= 0.12.0 

## Backup

Go to the directory where your Gogs binary is located, and execute following command:

```
./gogs backup
```

Without any flags, `backup` command will pack up all `gogs-repositories`, `data`, `custom` and database into a single zip archive (e.g. `gogs-backup-xxx.zip`) under the current directory.

It could be a bad idea if your `gogs-repositories` contains GB of raw data, in that case, you can apply `--exclude-repos` flag:

```
./gogs backup --exclude-repos
```

If your `custom/conf/app.ini` is somewhere unusual, make sure you specify it via `--config` flag like always:

```
./gogs backup --config=my/custom/conf/app.ini
```

### Database

If you're only interested in backup your database, or want to migrate from one database engine (e.g. SQLite3) to another engine (e.g. PostgreSQL), `--database-only` is your friend:

```
./gogs backup --database-only
```

The backup format of database are portable JSON files, each file corresponds to a database table, you can do whatever you want with those files.

## Restore

The `restore` command also has flags to indicate only restore database or everything in backup archive:

```
./gogs restore --database-only --from="gogs-backup-xxx.zip"
```

If a table that is not presented in backup archive, whatever in current database table will remain unchanged.

### Custom config file

There are 2 steps to determine which `custom/conf/app.ini` command uses:

1. Use the one you specified vis flag --config.
1. Use the one stored in backup archive.

If both steps failed, sorry, impossible to perform restore process.

**Comment by Loqova** [comment]
When running gogs inside a Docker container (named 'gogs' in my case) I use:

````
docker exec gogs bash -c 'ls /app/gogs | rm -f gogs-backup-*.zip && cd /app/gogs && USER=git ./gogs backup' && docker cp gogs:$(docker exec gogs bash -c 'readlink -f $(ls /app/gogs | grep gogs-backup-*.zip -)') . && md5sum $(docker exec gogs bash -c 'ls /app/gogs | grep gogs-backup-*.zip -') > $(docker exec gogs bash -c 'ls /app/gogs | grep gogs-backup-*.zip -').md5 && docker exec gogs bash -c 'ls /app/gogs | rm gogs-backup-*.zip'
````

This first removes old back-ups inside the container, then makes a new back-up inside the docker container, copies the file to the current location (outside the container) and removes the backup from within the container again. It also makes an MD5 hash of the made back-up.

Inspired by: https://github.com/gogs/gogs/issues/4339

**Reply by unknwon** [reply]
Thanks for sharing!

**Reply by Hser2bio** [reply]
thanks !

**Reply by amazo-spw** [reply]
Damn it, it took me 5 hours to finally find a solution. Thank you for sharing.

**Reply by bohdantrotsenko** [reply]
deleting old backups first will backfire one day

**Reply by jdweb71** [reply]
thanks !

**Comment by whisperdancer** [comment]
I ran restore command with success. I checked the db and see it has restored the users and it has restored the repositories. However, I cannot login to my account. It returns error that username or password is incorrect. I checked the user table and see the users, but the password field is blank. Perhaps that's the issue. Shouldn't the restore operation restore user passwords also. Hoping there is a fix for this. Kindly advise. Thank you.

**Reply by urko-b** [reply]
same here

**Reply by ZJ1998888** [reply]
请问你解决这个问题了嘛？

**Comment by whisperdancer** [comment]
Unfortunately, nobody on gogs offers to help so I ditched it. I moved over to gitea on docker and it's superior.

**Comment by zfrhv** [comment]
what i did for gogs helm chart:
in source pod:
```bash
kubectl exec -n gogs <pod name> -- bash -c "HOME=/data/git USER=git /app/gogs/gogs backup --target /backup"
kubectl cp -n gogs <pod name>:/backup/<file> ./<file>
```
in destination gogs (make sure its running 1/1):
```bash
kubectl cp -n gogs ./<file> <pod name>:/backup/<file>
kubectl exec -n gogs <pod name> -- bash -c "mkdir /data/tmp; HOME=/data/git USER=git /app/gogs/gogs restore --from /backup/<file> -t /data/tmp; chown -R git:git /data"
```

**Comment by MaxKingPor** [comment]
I get an error when I want to migrate sqlite3 to postgresql
```text
$ ./gogs backup
2024/12/24 14:10:47 [ INFO] Backup root directory: /tmp/gogs-backup-960043014
2024/12/24 14:10:47 [ INFO] Packing backup files to: gogs-backup-20241224141047.zip
2024/12/24 14:10:47 [ INFO] Dumping repositories in "/mnt/disk/apps/gogs/data/gogs-repositories"
2024/12/24 14:10:47 [ INFO] Repositories dumped to: /tmp/gogs-backup-960043014/repositories.zip
2024/12/24 14:10:48 [ INFO] Backup succeed! Archive is located at: gogs-backup-20241224141047.zip
$ ./gogs restore --from gogs-backup-20241224141047.zip
2024/12/24 14:14:59 [ INFO] Restoring backup from: gogs-backup-20241224141047.zip
2024/12/24 14:15:00 [FATAL] [....io/gogs/internal/cmd/restore.go:130 runRestore()] Failed to import 'custom': rename /tmp/gogs-backup/custom /mnt/disk/apps/gogs/custom: invalid cross-device link
```

**Comment by xavgru12** [comment]
This ignores --config in both backup and restore, no matte what I try
